src/chat.rs

```diff
@@ -815,12 +818,17 @@ pub fn send_msg(context: &Context, chat_id: u32, msg: &mut Message) -> Result<u3
         let forwards = msg.param.get(Param::PrepForwards);
         if let Some(forwards) = forwards {
             for forward in forwards.split(' ') {
-                let id: i32 = forward.parse().unwrap_or_default();
-                if 0 == id {
-                    // avoid hanging if user tampers with db
-                    break;
-                } else if let Ok(mut copy) = Message::load_from_db(context, id as u32) {
-                    send_msg(context, 0, &mut copy)?;
+                match forward
+                    .parse::<u32>()
+                    .map_err(|_| InvalidMsgId)
+                    .map(|id| MsgId::new(id))
```

Comment:
please impl FromStr for MessageId, then this can become much nicer

https://doc.rust-lang.org/std/str/trait.FromStr.html

Reply:
I considered that, but it breaks the notion that MsgId is opaque which I think is nice to aim for.  The cost of doing this thing here is not worth incorporating that into the API I think.  This really here is serialisation to/from Params.  I'd say we should have ToParams/FromParams traits for this at some point (but even that I'm not sure since I'd also consider putting the params directly into the db as columns in the sql).  In the mean time I can live with this.

Reply:
what do you mean with "opaque" in the first sentence? serializing to/from Str, if implemented in message.rs, can maintain opaqueness i'd think.